### ISSUE TITLE
New Feature: onZoneDone

### ISSUE THREAD

**Issue by robrbecker** [angular/zone.js#8]
tl;dr I would like an onZoneDone callback to be called once the function passed in to run is complete, and any tree of async code spawned from that function.

I would like to request a new feature.  When I first looked at zone.js I was under the impression that onZoneEnter and onZoneLeave "contained" the function that was invoked via run **and** any async code kicked off from that function. I discovered that instead a zone can enter and leave more than once. 

For Example:

```
zone.onZoneEnter = function() {
  console.log('entering');
};
zone.onZoneLeave = function() {
  console.log('leaving');
};
zone.onZoneDone = function() {
  console.log('All done finally');
};
zone.run(function() {
  console.log('hello');
  setTimeout(function() {
    console.log('this happened later');
  }, 3000);
});
```

This currently outputs:
entering
hello
leaving
leaving
entering
this happened later
leaving

I'd like it to output:
entering
hello
leaving
leaving
entering
this happened later
leaving
**All done finally**

You'll see that onZoneDone gets called after the last remaining setTimeout completes.

@btford
